kratos/conditions/penalty_coupling_condition.cpp

```diff
@@ -0,0 +1,221 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Tobias Tescheamacher
+//
+
+// System includes
+
+// External includes
+#include "conditions/penalty_coupling_condition.h"
+
+// Project includes
+
+namespace Kratos
+{
+
+    void PenaltyCouplingCondition::CalculateAll(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag)
+    {
+        KRATOS_TRY
+
+        DofVariablesContainer const& r_dof_variables = this->GetProperties().GetValue(DOF_VARIABLES);
+
+        const double penalty = GetProperties()[INITIAL_PENALTY];
+
+        const auto& r_geometry_master = GetGeometry().GetGeometryPart(0);
+        const auto& r_geometry_slave = GetGeometry().GetGeometryPart(1);
+
+        // Size definitions
+        const int number_of_nodes_master = r_geometry_master.size();
+        const int number_of_nodes_slave = r_geometry_slave.size();
+
+        const int mat_size = r_dof_variables.size() * (number_of_nodes_master + number_of_nodes_slave);
+
+        // Memory allocation
+        if (CalculateStiffnessMatrixFlag) {
+            if (rLeftHandSideMatrix.size1() != mat_size) {
+                rLeftHandSideMatrix.resize(mat_size, mat_size, false);
+            }
+            noalias(rLeftHandSideMatrix) = ZeroMatrix(mat_size, mat_size);
+        }
+        if (CalculateResidualVectorFlag) {
+            if (rRightHandSideVector.size() != mat_size) {
+                rRightHandSideVector.resize(mat_size, false);
+            }
+            rRightHandSideVector = ZeroVector(mat_size);
+        }
+
+        // Integration
+        const GeometryType::IntegrationPointsArrayType& integration_points = r_geometry_master.IntegrationPoints();
+        for (IndexType point_number = 0; point_number < integration_points.size(); point_number++)
+        {
+            Matrix N_master = r_geometry_master.ShapeFunctionsValues();
```

Comment:
const, and if it is something store & too
```suggestion
            const Matrix N_master = r_geometry_master.ShapeFunctionsValues();
```